version: "3.5"

services:
  jupyterhub:
    restart: on-failure
    image: {{docker_illumidesk_jhub_image}}
    depends_on:
      - reverse-proxy
      - jupyterhub-db
      - setup-course
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:rw
      - {{mnt_root}}/{{org_name}}/home:/home
      - data:/data
      - ./reverse-proxy:/etc/traefik
    env_file:
      - env.jhub
    command: >
      jupyterhub -f /etc/jupyterhub/jupyterhub_config.py --debug
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.jhub_router.service=jupyterhub"
  jupyterhub-db:
    image: postgres:11.6-alpine
    container_name: jupyterhub-db
    restart: always
    env_file:
      - env.jhub
    volumes:
      - db:/var/lib/postgresql/data
  reverse-proxy:
    image: traefik:v1.7-alpine
    container_name: reverse-proxy
    ports:
      - "8000:8000"
      - "443:443"
      - "8080:8080"
    volumes:
      - ./reverse-proxy/:/etc/traefik/
      - /var/run/docker.sock:/var/run/docker.sock
    restart: on-failure    
  {{course_id}}:
    restart: on-failure
    image: {{docker_illumidesk_notebook_grader_image}}
    container_name: {{course_id}}
    user: root
    volumes:
      - {{mnt_root}}/{{org_name}}/home/grader-{{course_id}}:/home/grader-{{course_id}}
      - {{mnt_root}}/{{org_name}}/exchange:/srv/nbgrader/exchange
    env_file:
      - env.service
    command: >
      start-notebook.sh --group=formgrade-{{course_id}} --allow-root
  setup-course:
    restart: on-failure
    image: {{docker_setup_course_image}}
    container_name: setup-course
    command: hypercorn --bind 0.0.0.0:8000 --workers 1 illumidesk.setup_course.app:app
    env_file:
      - env.setup_course
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:rw
      - {{mnt_root}}:{{mnt_root}}
      - data:/srv/jupyterhub
      - {{ working_dir }}:{{ working_dir }}

volumes:
  data:
    external:
      name: jupyterhub-data
  db:
    external:
      name: db-data
  
networks:
  default:
    external:
      name: jupyter-network
